### PR TITLE
fix mainwindow and preferences window positioning

### PR DESCRIPTION
**Pull request by JosepMaJAZ**
fix positioning when reloading skin from developer menu. Also fixed Preferences window positioning

### CODE REVIEW

**Comment by sblaisot** [comment]
I assigned bug https://bugs.launchpad.net/mixxx/+bug/1533575 to you.

**Comment by JosepMaJAZ** [comment]
Ok. I will also fix that later today.

**Comment by sblaisot** [comment]
I thought i was fixed by this PR already.
I will try your PR and let you know.

**Comment by JosepMaJAZ** [comment]
I fixed what was causing it to go outside of the screen, but I haven't
added yet a "move it back if it is already outside". But that is
implemented in the other correction in mixxx.cpp, so it is simply a
copy&paste missing.

**Comment by sblaisot** [comment]
I just tested using Appveyor artifact and moving outside of screen doesn't seems to be fixed.
Each time I show preference window, it moves up&left by a (probably) fixed amount of pixel. Showing preference window several times finally move it outside of screen.

**Comment by JosepMaJAZ** [comment]
On which OS? I tested it on linux.

**Comment by sblaisot** [comment]
On windows

**Comment by JosepMaJAZ** [comment]
Ok, so far, I can confirm that on windows the solution didn't work, and at some point, it stopped working on linux too, but now I've got both mostly working, although with a different code path for each one. I'll commit it after a couple more tests

File: src/preferences/dialog/dlgpreferences.cpp [hunk 1]
```diff
@@ -368,15 +368,38 @@ void DlgPreferences::onShow() {
...
+#if defined(Q_OS_WIN)
+    resize(m_geometry[2].toInt(), m_geometry[3].toInt());
+#else 
+    // On linux, when the window is opened for the first time by the window manager, QT does not have
+    // information about the frame size so the offset is zero. As such, the first time it opens the window
+    // does not include the offset, so it is moved from the last position it had.
+    // Excluding the offset from the saved value tries to fix that.
+    int offsetX = geometry().left() - frameGeometry().left();
+    int offsetY = geometry().top() - frameGeometry().top();
+    newX+=offsetX;
```
Comment (Be-ing) [inline src/preferences/dialog/dlgpreferences.cpp]: nitpick: please use spaces around the operator.

File: src/preferences/dialog/dlgpreferences.cpp [hunk 2]
```diff
@@ -367,12 +368,38 @@ void DlgPreferences::onShow() {
...
+//    int offsetX = geometry().left() - frameGeometry().left();
+//    int offsetY = geometry().top() - frameGeometry().top();
+//    qDebug() << "ONSHOW framegeo " << frameGeometry() << " geo " << geometry() << " mgeometry " << m_geometry;
+//    qDebug() << "ONSHOW offsets " << offsetX << " " << offsetY;
+//    qDebug() << "ONSHOW  newxy" << newX << " " << newY;
 
     // Update geometry with last values
-    setGeometry(m_geometry[0].toInt(),  // x position
-                m_geometry[1].toInt(),  // y position
+#if defined(Q_OS_WIN)
```
Comment (sblaisot) [inline src/preferences/dialog/dlgpreferences.cpp]: You should probably use `#ifdef __WINDOWS__` for code consistency

File: src/preferences/dialog/dlgpreferences.cpp [hunk 3]
```diff
@@ -463,8 +490,21 @@ void DlgPreferences::switchToPage(DlgPreferencePage* pWidget) {
...
 void DlgPreferences::moveEvent(QMoveEvent* e) {
     if (m_geometry.length() == 4) {
-        m_geometry[0] = QString::number(e->pos().x());
-        m_geometry[1] = QString::number(e->pos().y());
+        // Warning! geometry does NOT include the frame/title.
+//        int offsetX = geometry().left() - frameGeometry().left();
+//        int offsetY = geometry().top() - frameGeometry().top();
+//        qDebug() << "MOVEEVENT framegeo " << frameGeometry() << " geo " << geometry() << " epos " << e->pos();
+//        qDebug() << "MOVEEVENT offsets " << offsetX << " " << offsetY;
+#if defined(Q_OS_WIN)
```
Comment (sblaisot) [inline src/preferences/dialog/dlgpreferences.cpp]: You should probably use `#ifdef __WINDOWS__` for code consistency

File: src/mixxx.cpp [hunk 4]
```diff
@@ -1068,7 +1068,9 @@ void MixxxMainWindow::slotViewFullScreen(bool toggle) {
 }
 
 void MixxxMainWindow::slotOptionsPreferences() {
-    m_pPrefDlg->setHidden(false);
+//    m_pPrefDlg->setHidden(false);
```
Comment (sblaisot) [inline src/mixxx.cpp]: Remove dead code

File: src/mixxx.cpp [hunk 5]
```diff
@@ -1178,8 +1183,15 @@ void MixxxMainWindow::rebootMixxxView() {
...
-        move(initPosition.x() + (initSize.width() - m_pWidgetParent->width()) / 2,
-             initPosition.y() + (initSize.height() - m_pWidgetParent->height()) / 2);
+        // Not all OSs and/or window managers keep the window inside of the screen, so force it.
+        int newX = initPosition.x() + (initSize.width() - m_pWidgetParent->width()) / 2;
+        int newY = initPosition.y() + (initSize.height() - m_pWidgetParent->height()) / 2;
+        newX = std::max(0, std::min(newX, QApplication::desktop()->screenGeometry().width()- m_pWidgetParent->width()));
+        newY = std::max(0, std::min(newY, QApplication::desktop()->screenGeometry().height() - m_pWidgetParent->height()));
+//        qDebug() << "rebootMixxxView old pos:" << initPosition << " old size:" << initSize;
+//        qDebug() << "new widdget width/height:" << m_pWidgetParent->width() << m_pWidgetParent->height();
+//        qDebug() << "new x/y target:" << newX << "," << newY;
```
Comment (sblaisot) [inline src/mixxx.cpp]: Remove debugging code before merge

File: src/preferences/dialog/dlgpreferences.cpp [hunk 6]
```diff
@@ -367,12 +368,38 @@ void DlgPreferences::onShow() {
...
+    newX = std::max(0, std::min(newX, QApplication::desktop()->screenGeometry().width()- m_geometry[2].toInt()));
+    newY = std::max(0, std::min(newY, QApplication::desktop()->screenGeometry().height() - m_geometry[3].toInt()));
+    m_geometry[0] = QString::number(newX);
+    m_geometry[1] = QString::number(newY);
+    
+//    int offsetX = geometry().left() - frameGeometry().left();
+//    int offsetY = geometry().top() - frameGeometry().top();
+//    qDebug() << "ONSHOW framegeo " << frameGeometry() << " geo " << geometry() << " mgeometry " << m_geometry;
+//    qDebug() << "ONSHOW offsets " << offsetX << " " << offsetY;
+//    qDebug() << "ONSHOW  newxy" << newX << " " << newY;
```
Comment (sblaisot) [inline src/preferences/dialog/dlgpreferences.cpp]: Remove debugging code before merge

File: src/preferences/dialog/dlgpreferences.cpp [hunk 7]
```diff
@@ -463,8 +490,21 @@ void DlgPreferences::switchToPage(DlgPreferencePage* pWidget) {
 
 void DlgPreferences::moveEvent(QMoveEvent* e) {
     if (m_geometry.length() == 4) {
-        m_geometry[0] = QString::number(e->pos().x());
-        m_geometry[1] = QString::number(e->pos().y());
+        // Warning! geometry does NOT include the frame/title.
+//        int offsetX = geometry().left() - frameGeometry().left();
+//        int offsetY = geometry().top() - frameGeometry().top();
+//        qDebug() << "MOVEEVENT framegeo " << frameGeometry() << " geo " << geometry() << " epos " << e->pos();
+//        qDebug() << "MOVEEVENT offsets " << offsetX << " " << offsetY;
```
Comment (sblaisot) [inline src/preferences/dialog/dlgpreferences.cpp]: Remove debugging code before merge

File: src/preferences/dialog/dlgpreferences.cpp [hunk 8]
```diff
@@ -463,8 +490,21 @@ void DlgPreferences::switchToPage(DlgPreferencePage* pWidget) {
...
+#if defined(Q_OS_WIN)
+    Q_UNUSED(e);
+        m_geometry[0] = QString::number(frameGeometry().left());
```
Comment (sblaisot) [inline src/preferences/dialog/dlgpreferences.cpp]: sapce vs tab identation mismatch with previous line ?

**Review by sblaisot — CHANGES_REQUESTED** [review]
Thank you so much for taking care of this ! I tested this version and it works well. No more preference window moving outside of screen, fixed and predictible position.
I left you somme minor comments.
Other than that, LGTM.

**Comment by daschuer** [comment]
Thank you very much for fixing this one! :-)

**Comment by daschuer** [comment]
Just noticed one remaining thing: 
On Linux the Mixxx window is still moving slowly right after every restart. This PR fixes th case that the window is moving out of the visible area. Thats good.
Do you have an idea why the moving happens in the first place?

**Comment by JosepMaJAZ** [comment]
That has to do with the way X window and the windows manager works. I added
a way to compensate for the possibility of it being erroneous at first
open, but if it isn't working for you, it might not be generic enough.

**Pull request merged**